dashboard_server.py

```
@@ -0,0 +1,615 @@
+        if not path.exists():
+            return []
+        try:
+            df = pd.read_csv(path)
+            df = df.where(pd.notna(df), None)
+            return df.to_dict('records')
+        except FileNotFoundError:
+            # File doesn't exist - handled by path.exists() check above, but include for completeness
+            return []
+        except PermissionError as e:
+            import logging
+            logging.getLogger(__name__).error(f"Permission denied reading {path}: {e}")
```

Ruff (0.12.2)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)